### PR TITLE
[sphinx] No more undocumented objects.

### LINKED ISSUE #7602: Plan to get rid of objects without body
#7556 added a new warning for objects without a body (generally this is due to the body being incorrectly indented and not recognized as such). Cf. https://github.com/coq/coq/tree/master/doc/sphinx#improper-nesting.
It also introduced the `:undocumented:` flag to put in place of a body when needed (cf. https://github.com/coq/coq/tree/master/doc/sphinx#coq-objects).

However, this warning is currently turned off by default because there are too many occurrences and we cannot address them in one day. What we could do would be to address this warning chapter by chapter and to turn the warning on only for chapters that are free of it (to make sure that it doesn't come back). This would require enhancing the configuration option added by #7556 to allow a finer-grained configuration (chapter by chapter). Any help on this topic from any volunteer would be welcome.

### PR DESCRIPTION
**Pull request by Zimmi48**
**Kind:** documentation fix

Fixes / closes #7602 

@cpitclaudel The change in the `coqrst` code remove the requirement for "documentation" for `prodn` objects.

@gares The SSReflect chapter is always the hardest to fix to make it closer to Sphinx requirements because it is generally written in a style that's not the expected one for this manual. I did what I could to reformulate in a way that is closer to what is expected (please check) and when I really couldn't without changing too much, I just put an `:undocumented:` option instead, but longer term these options should go away.

### CODE REVIEW

**Review by cpitclaudel — APPROVED** [review]
LGTM, though I'm not qualified wrt the SSReflect chapter

**Comment by Zimmi48** [comment]
Let's give a few days to @gares to react before merging. @amahboubi or any member of @coq/ssreflect-maintainers could be alternative reviewers.

**Comment by Zimmi48** [comment]
No comment after more than two weeks. Let's not delay this merge further. I'll rebase first just to be sure that no new warning has been introduced in the meantime.

**Comment by Zimmi48** [comment]
Good. No new warning. The CI failure is irrelevant: math-comp is currently broken.

**Comment by cpitclaudel** [comment]
Thanks!

**Pull request merged**
